File: merge-sort.rb
def merge(left_side, right_side)
  sorted_arr = []
  until left_side.empty? || right_side.empty?
    if left_side.first <= right_side.first
      sorted_arr << left_side.shift
    else
      sorted_arr << right_side.shift
    end
  end
  sorted_arr.concat(left_side).concat(right_side)
end

def merge_sort(arr)

  if arr.length <= 1
    return arr
  end

  mid_index = arr.length / 2
  left_side = arr[0...mid_index]
  right_side = arr[mid_index...arr.length]

  merge(merge_sort(left_side), merge_sort(right_side))
end

File: mergeSort.js

// Works

function mergeSort(arr) {
  if (arr.length <= 1) {
    return arr
  }
  let midpoint = Math.floor(arr.length/2)
  let leftSide = arr.slice(0, midpoint) // note I am passing these below by reference, and they get widdled away into new arr.
  let rightSide = arr.slice(midpoint)
  return merge(mergeSort(leftSide), mergeSort(rightSide))
}

function merge(leftSide, rightSide) {
  let mergedArr = []

  while (leftSide.length != 0 && rightSide.length != 0) {
    if (leftSide[0] < rightSide[0]) {
      mergedArr.push(leftSide.shift())
    } else {
      mergedArr.push(rightSide.shift())
    }
  }

  return mergedArr.concat(leftSide).concat(rightSide)
}
